daffodil-runtime1/src/main/scala/org/apache/daffodil/dpath/FNFunctions.scala

```diff
@@ -803,6 +805,85 @@ case class FNFloor(recipe: CompiledDPath, argType: NodeInfo.Kind) extends FNOneA
   }
 }
 
+/**
+ * This is "arithmetic" shift left, meaning it sign-extends a negative signed value.
+ * dfdl:leftShift for signed values does sign-extension, and if you want logical (zero-filling) left shift you must use an unsigned type
+ *
+ * @param recipes
+ * @param argType
+ */
+case class DFDLXLeftShift(recipes: List[CompiledDPath], argType: NodeInfo.Kind) extends FNTwoArgs(recipes)  {
+  override def computeValue(arg1: DataValuePrimitive, arg2: DataValuePrimitive, dstate: DState): DataValuePrimitive = {
+
+    val shiftLong = arg2.getLong
+    val shift = shiftLong.toInt
+    val width = argType.asInstanceOf[PrimNumeric].width.get
+    Assert.invariant(shift >= 0)
+    if(shift>=width)
+      dstate.SDE("Left shift not supported for shift greater or equal to "+ width)
+
+    argType match {
+      case NodeInfo.Long => arg1.getLong << shift
+
+      case NodeInfo.Int => arg1.getInt << shift
+
+      case NodeInfo.Short => (arg1.getShort << shift).toShort
+
+      case NodeInfo.Byte => (arg1.getByte << shift).toByte
+
+      case NodeInfo.UnsignedLong =>arg1.getBigInt.shiftLeft(shift)
+
+      case NodeInfo.UnsignedInt => ULong(arg1.getLong << shift).toLong
+
+      case NodeInfo.UnsignedShort => UInt(arg1.getInt << shift).toInt
+
+      case NodeInfo.UnsignedByte => UInt(arg1.getShort << shift).toShort
```

Comment:
I think these unsigned shifts are still incorrect. For example, if you create a test where you shift the maximum value of the unsigned type one bit left, I think you will get the wrong answer.

As a specific example, the max value of unsigned byte is 255. If we shift that left we should get 254--the left significant bit becomes 0 and the most significant bit is masked off after the shift:
```
76543210 // bit position
--------
11111111 // = 255
11111110 // = 255 << 1 = 254
```
But if we run the following, which is equivalent to your code with a unsignedByte value of 255:
```scala
scala> passera.unsigned.UInt(255.toShort << 1).toShort
res1: Short = 510
```
Then the result is 510, which is ``0b1000000000`` in binary. Which is not correct. That's more than 8 binary digits, which isn't right for an unsigned byte.

For UnsignedByte as an example, the correct shift would be:
```scala
case NodeInfo.UnsignedByte => (UByte(arg1.getShort.toByte) << 1).toUByte.toShort
```
Note that the ``toUByte`` is needed because ``UByte << 1`` returns a ``UInt``. The ``.toUByte`` will essentially mask off high order bits.

I think maybe an alternative approach is to use UInt for everything and then use toUByte (or the appropriate usngiend type). For example:

```scala
case NodeInfo.UnsignedByte => UInt(arg1.getShort << shift).toUByte.toShort
```
So we first get the arg1 value, which is a Short in he case of UnsignedByte. We shift that left the shift value. This returns an Int, which can create a value too large for UnsigneByte. Since the shift returns an Int, we can simply convert that to an UInt. And then we can mask off those excess bits by calling ``.toUByte``. And finally we convert it to a Short, which the infoset requires.

I think those two options are functionally the same?

Regardless, we really need to add tests for these edge cases to ensure masking is correct. For example, for each unsigned type, we need a test that takes the max value and shifts it left 1 and makes sure the result is one less. We might also want a test that takes the max value and shifts left the max shift (i.e. ``width - 1``) and makes sure the result is ``2^(width - 1)``.

Reply:
@stevedlawrence For byte and short I have made the changes as mentioned by you.I have added test cases by shifting with max value for unsigned value, they work as expected.

Reply:
@Samarth08, you've get tests for shifting max width, but not tests for shifting max value. For example, xs:unsignedByte should have a test with data "255,1" so we shift the value 255 left 1 bit. xs:unsignedShort should have a test for "65535,1", and so on, up until xs:unsignedLong, which should have a test for "18446744073709551615,1". I'm not entirely sure if what we have is correct for xs:unsignedLong and xs:unsingedInt. Tests that have these values should give a little more confidence.

For example, I think UnsignedInt wants to be:
```scala
case UnsignedInt => ULong(arg1.getLong << shift).toUInt.toLong
```
The toUInt is important to mask off those extra bits. Extra tests will ensure this is correct.

Likewise, taking these same max values and shifting the max width would be useful tests as well to make sure we're completely masking off all bits.

Reply:
How do we handle unsigned long in this case 18446744073709551615,1, since we are using BigInt for unsigned long.

Reply:
I'm actually having trouble trying to figure out the right thing to do for BigInteger. Using ULong, I'm not getting the expected results. For example, I would think we could do somthing like this:

```scala
case UnsignedLong => (ULong(arg1.getBigInt.longValue()) << shift).toBigInt
```
BigInt.longValue should get us a value that we can pass into ULong. The ULong shift left should mask things correctly, and the converting to a BigInt should work fine.

But it's not working as I'd expect, and now I'm wondering if there's a bug in ULong?

For example, see the following:
```scala
scala> val max = passera.unsigned.ULong.MaxValue
max: passera.unsigned.ULong = 9223372036854775808

scala> val hex = max.toHexString
hex: String = ffffffffffffffff

scala> val shiftOne = max << 1
shiftOne: passera.unsigned.ULong = 9223372036854775807

scala> val shiftOneHex = shiftOne.toHexString
shiftOneHex: String = fffffffffffffffe
```
So the hex values are correct, and the shiftOne value is one less than the max value, but the toString values are wrong. Not sure yet what's wrong with ULong, but seems like the conversion to a string is wrong. And the conversion to a string can use a BigInt, so maybe there is a bug in how we convert to a BigInt?

Reply:
I think maybe the problem is related to this line in ULong.scala:
```scala
val MaxValueAsBigInt = JBigInt.valueOf(Long.MinValue).abs
```
The max value of a ULong is not the absolute value of the smallest Long. That gives a value of 9223372036854775808, which is about half as big as the max ULong value, which is 18446744073709551615. And that MaxValueAsBigInt variable is used in toBigInt and toString calculations.

If I change that line to 
```scala
val MaxValueAsBigInt = new JBigInt("18446744073709551615")
```
Then my shift examples above work as expected. There is one passera test that fails, but it's doing the same abs long value thing, which I think is wrong.

@mbeckerle, any ideas? I could use a second opinion to make sure I'm thinking about this all correctly.

This is subtle edge case with a probably rarely used part of code, so maybe we just haven't had ULong's get big enough to hit this? Or maybe I'm missing something....

Reply:
I concur that in ULong.scala, this line looks incorrect:
```
val MaxValueAsBigInt = JBigInt.valueOf(Long.MinValue).abs
```
That's just simply incorrect. That's 0x8000000000000000. 

There are other mistakes also in toString and toBigInt that are due to this error.

I created https://issues.apache.org/jira/browse/DAFFODIL-2522 for this, and as part of investigating, have a PR of the fixes. 

https://github.com/apache/daffodil/pull/571

The max value of ULong should be:
```
val MaxValueAsBigInt = new JBigInt("FFFFFFFFFFFFFFFF", 16)
```
Using hex there just to be clear this is 64 bits of 1. 

With that correction, ULong.MaxValue.toBigInt works correctly, producing 18446744073709551615.
But there are still wierd things. 
E.g.,
```
scala> ULong(Long.MinValue)
res16: passera.unsigned.ULong = 18446744073709551615
``` 
huh? Not even close. 
That should be the same as java.math.BigInteger("8000000000000000", 16) or 9223372036854775808

So I fixed these issues and added tests for ULong.

Reply:
I have made changes for unsigned long, but not added test cases for MAX value.Do I have to take a pull after the PR is merged.Then add the test cases or it is good without the pull.
Also Should I start with bitwise OR,ADD,NOT?

Reply:
I suggest waiting for #571 to be merged, and then rebase + squash this branch on top of that with added test.

OR/AND/NOT are very reasonable next steps. Though, I'd perfer we do that as a new pull request just so we can get this one merged and closed.